Storage/SQL/SQLiteHistory.swift

```diff
@@ -105,6 +148,214 @@ open class SQLiteHistory {
+        }
+    }
+
+    func getSites(whereURLContains filter: String?, historyLimit limit: Int, bookmarksLimit: Int) -> Deferred<Maybe<Cursor<Site>>> {
+        let factory = SQLiteHistory.basicHistoryColumnFactory
+
+        let params = FrecencyParams.urlCompletion(bookmarksLimit: bookmarksLimit, whereURLContains: filter ?? "", groupClause: "GROUP BY historyID ")
+        let (query, args) = getFrecencyQuery(historyLimit: limit, params: params)
+
+        return db.runQuery(query, args: args, factory: factory)
+    }
+
+    func updateTopSitesCacheQuery() -> (String, Args?) {
+        let limit = Int(prefs.intForKey(PrefsKeys.KeyTopSitesCacheSize) ?? TopSiteCacheSize)
+
+        let (whereData, groupBy) = topSiteClauses()
+        let params = FrecencyParams.topSites(groupClause: groupBy, whereData: whereData)
+        let (frecencyQuery, args) = getFrecencyQuery(historyLimit: limit, params: params)
+
+        // We must project, because we get bookmarks in these results.
+        let insertQuery = [
+            "WITH siteFrecency AS (\(frecencyQuery))",
+            "INSERT INTO \(TableCachedTopSites)",
+            "SELECT historyID, url, title, guid, domain_id, domain,",
+            "localVisitDate, remoteVisitDate, localVisitCount, remoteVisitCount,",
+            "iconID, iconURL, iconDate, iconWidth, frecencies",
+            "FROM (SELECT * FROM",
+            "siteFrecency LEFT JOIN view_history_id_favicon ON",
+            "siteFrecency.historyID = view_history_id_favicon.id)"
+            ].joined(separator: " ")
+
+        // @TODO: remove the LEFT JOIN to fill in the icon columns, it is just there because the code has been doing this historically.
+        // Those favicon data columns are not used, and view_history_id_favicon appears to only contain null data.
+
+        return (insertQuery, args)
+    }
+
+    private func topSiteClauses() -> (String, String) {
+        let whereData = "(\(TableDomains).showOnTopSites IS 1) AND (\(TableDomains).domain NOT LIKE 'r.%') AND (\(TableDomains).domain NOT LIKE 'google.%') "
+        let groupBy = "GROUP BY domain_id "
+        return (whereData, groupBy)
+    }
+
+
+    enum FrecencyParams {
+        case urlCompletion(bookmarksLimit: Int, whereURLContains: String, groupClause: String)
+        case topSites(groupClause: String, whereData: String)
+    }
+
+    private func getFrecencyQuery(historyLimit: Int, params: FrecencyParams) -> (String, Args?) {
+        let bookmarksLimit: Int
+        let groupClause: String
+        let whereData: String?
+        let urlFilter: String?
+
+        switch params {
+        case let .urlCompletion(bmLimit, filter, group):
+            bookmarksLimit = bmLimit
+            urlFilter = filter
+            groupClause = group
+            whereData = nil
+        case let .topSites(group, whereArg):
+            groupClause = group
+            whereData = whereArg
+            bookmarksLimit = 0
+            urlFilter = nil
```

Comment:
Urgh, put these in the same order in each branch!

But even better: implement a private method on `FrecencyParams` that returns `(bookmarksLimit, groupClause, whereData, urlFilter)`, and just do

```
let (bookmarksLimit, groupClause, whereData, urlFilter) = params.fooBar();
```

Reply:
Fixed order.

Not agreeing about tuple of params though, in this case I don't think the added variable setter abstraction here will reduce cognitive load reading this complex function, as referring to this block seems easiest to read and refer back to while reading the func.